### ISSUE TITLE
Does Meson version number follow semver?

### ISSUE THREAD

**Issue by groengpx** [mesonbuild/meson#11261]
Is it sort of guaranteed that if my project builds with meson version 1.0.0, that it will also build with any 1.x.y version? (of course, until we hit 2.0.0)

**Comment by eli-schwartz** [comment]
Policy up until now -- for the pre-1.0 versions of Meson -- has already been fairly strict.

- if Meson's documentation says something should work, it should work
- if something is documented to work, we can't remove support for it if we decide it's bad, so instead we deprecate it and emit a versioned deprecation message
- if something is not documented to work, things got a bit fuzzy.
  - if it's a matter of missing typechecking on functions, we usually considered it a bug fix, not a breaking change -- actually we aren't even usually sure what happens if you violate the typechecking, most of the time it simply explodes with a python traceback and we didn't exhaustively check. But, sometimes we later discovered it was silently ignored. If several projects turn out to be relying on it by accident, we would usually turn it into a deprecation warning for a release or two and give them time to fix things.
  - in one case, the systemd project relied on implicit type casting from bool to int, which breaks our syntax guarantee that we never cast types and do the wrong thing with them. In that case it turned out to be only one project and they fixed it on their end anyway. We didn't do anything there.
  - other changes still usually get a deprecation warning, whatever the cause


As for post-1.0 policy, fairly certain the idea was to indeed treat things as semver and provide much firmer stability guarantees, this is one of the reasons I wanted to wait until typechecking for build targets was finished. :(

If your project builds with Meson 1.0.0 without deprecation warnings, then it *should* build with any 1.x.y version as well -- and if it does not do so, that's a Meson bug as far as I'm concerned, regardless of the reason. (Unfortunately I cannot guarantee that Meson will never have a bug :) only that we try hard to avoid that from happening.)

And if you do get deprecation warnings, you'll have at least until the next semver major version after the warning first showed up, but we haven't really discussed what to do after that. We also need to keep in mind, that we really don't want to remove functionality that is still required to accomplish something on versions of Meson shipped in popular stable distros. There's two obvious potential approaches:
- once something is deprecated in an X.0.0 version, it can be removed in a minor release, once the ancient versions that it's backwards compatible with aren't in use anywhere, anymore
- remove things only during an X.0.0 release, but if it's deprecated too close to that release, wait until (X+1).0.0

Whatever we do, we *do* need to document what our expectations are (both in the HTML docs and, I think, in the warning messages themselves). So this ticket seems like a good place to track that documentation.

**Comment by dcbaker** [comment]
Yeah, it's pretty unfortunate the type checking stuff didn't get done, because there's a *ton* of UB in the build targets especially. It's also unfortunate that we didn't remove any of the long deprecated functionality (stuff that was out in 0.3x and 0.4x especially, since now we're stuck with all of that stuff till 2.0 :/

**Comment by bonzini** [comment]
Closing - since 1.0, it does.
